spec/compiler/semantic/warnings_spec.cr

```diff
@@ -308,6 +308,123 @@ describe "Semantic: warnings" do
+    it "warns when a deprecated positional argument is passed" do
+      assert_warning <<-CRYSTAL,
+        def foo(a, @[Deprecated] b, c)
+        end
+
+        foo(1, 2, 3)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument b."
+    end
+
+    it "warns when a deprecated keyword argument is passed" do
+      assert_warning <<-CRYSTAL,
+        def foo(x, *, a, @[Deprecated] b)
+        end
+
+        foo(0, a: 1, b: 2)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument b."
+    end
+
+    it "warns when a deprecated splat argument is passed" do
+      assert_warning <<-CRYSTAL,
+        def foo(a, @[Deprecated] *args)
+        end
+
+        foo(1, 2)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument args."
+    end
+
+    it "warns when a deprecated double splat argument is passed" do
+      assert_warning <<-CRYSTAL,
+        def foo(*, a, @[Deprecated] **opts)
+        end
+
+        foo(a: 1, bad: 2)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument opts."
+    end
+
+    it "warns when a deprecated default positional argument is passed" do
+      assert_warning <<-CRYSTAL,
+        def foo(a, @[Deprecated] b = nil, c = nil)
+        end
+
+        foo(1, 2)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument b."
+    end
+
+    it "warns when a deprecated default keyword argument is passed" do
+      assert_warning <<-CRYSTAL,
+        def foo(*, a, @[Deprecated] b = nil)
+        end
+
+        foo(a: 1, b: 2)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument b."
+    end
+
+    it "doesn't warn when a deprecated default positional argument isn't explicitly passed" do
+      assert_no_warning <<-CRYSTAL
+        def foo(a, @[Deprecated] b = nil, c = nil)
+        end
+
+        foo(1)
+        foo(1, c: 3)
+      CRYSTAL
+    end
+
+    it "doesn't warn when a deprecated default keyword argument isn't explicitly passed" do
+      assert_no_warning <<-CRYSTAL
+        def foo(*, a, @[Deprecated] b = nil)
+        end
+
+        foo(a: 1)
+      CRYSTAL
+    end
+
+    it "warns when a default value calls a method with a deprecated arg" do
+      assert_warning <<-CRYSTAL,
+        def bar(@[Deprecated] x)
+        end
+
+        def foo(a, @[Deprecated] b = bar(a))
+        end
+
+        foo(1)
+      CRYSTAL
+      "warning in line 4\nWarning: Deprecated argument x."
+
+      assert_warning <<-CRYSTAL,
+        def bar(@[Deprecated] x)
+        end
+
+        def foo(a, @[Deprecated] b = bar(a))
+        end
+
+        foo(1, 2)
+      CRYSTAL
+      "warning in line 7\nWarning: Deprecated argument b."
+    end
+
+    pending "doesn't warn when default value calls method with deprecated arg using a deprecated arg" do
+      assert_no_warning <<-CRYSTAL
+        def bar(@[Deprecated] x)
+        end
+
+        def foo(@[Deprecated] a = 1, @[Deprecated] b = bar(a))
+        end
+
+        foo
+      CRYSTAL
```

Comment:
This is one hell of an edge-case: since `a` is deprecated, I don't think the `bar(a)` call should trigger a warning when it uses `a` for its default value... but maybe it should.

Disabled because it does complain about "Deprecated argument x.".

Reply:
It seems actually fine to me that this only complains about deprecated argument `x`.

`foo` is called without arguments, so that's fine. It brings the default values into play. And one of them calls `bar` with a deprecated argument.

Reply:
The example was too convoluted, it should be simplified:

```crystal
def bar(@[Deprecated] x)
end

def foo(@[Deprecated] b = bar(1))
end
```

And then it falls under the same "don't warn about deprecations in deprecated stuff" logic, here becoming "don't warn about deprecated call/arg in deprecated arg (default value)".